pkg/asset/cluster/openstack/securitygroups.go

```diff
@@ -0,0 +1,217 @@
+		for ipVersion, anyIP := range map[rules.RuleEtherType][]string{
+			rules.EtherType4: []string{"0.0.0.0/0"},
+			rules.EtherType6: func() []string {
+				switch len(machineV6CIDRs) {
+				case 0:
+					return []string{}
+				default:
+					return []string{"::/0"}
+				}
+			}(),
+		} {
+			addRules(ctx, r, masterGroup.ID, serviceAPI, ipVersion, anyIP)
+			addRules(ctx, r, masterGroup.ID, serviceICMP, ipVersion, anyIP)
+			addRules(ctx, r, workerGroup.ID, serviceICMP, ipVersion, anyIP)
+			addRules(ctx, r, workerGroup.ID, serviceHTTP, ipVersion, anyIP)
+			addRules(ctx, r, workerGroup.ID, serviceHTTPS, ipVersion, anyIP)
+			if mastersSchedulable {
+				addRules(ctx, r, masterGroup.ID, serviceHTTP, ipVersion, anyIP)
+				addRules(ctx, r, masterGroup.ID, serviceHTTPS, ipVersion, anyIP)
+			}
+		}
+
+		// In this loop: security groups with the machine CIDR as remote IPs
+		for ipVersion, CIDRs := range map[rules.RuleEtherType][]string{
+			rules.EtherType4: machineV4CIDRs,
+			rules.EtherType6: machineV6CIDRs,
+		} {
+			// In this loop: rules that equally apply to masters and workers
+			for _, groupID := range [...]string{masterGroup.ID, workerGroup.ID} {
+				addRules(ctx, r, groupID, serviceSSH, ipVersion, CIDRs)
+				addRules(ctx, r, groupID, serviceVXLAN, ipVersion, CIDRs)
+				addRules(ctx, r, groupID, serviceGeneve, ipVersion, CIDRs)
+				addRules(ctx, r, groupID, serviceIKE, ipVersion, CIDRs)
+				addRules(ctx, r, groupID, serviceESP, ipVersion, CIDRs)
+				addRules(ctx, r, groupID, serviceInternal, ipVersion, CIDRs)
+				addRules(ctx, r, groupID, serviceKubelet, ipVersion, CIDRs)
+				addRules(ctx, r, groupID, serviceNodeport, ipVersion, CIDRs)
+				addRules(ctx, r, groupID, serviceVRRP, ipVersion, CIDRs)
+			}
+
+			addRules(ctx, r, masterGroup.ID, serviceMCS, ipVersion, CIDRs)
+			addRules(ctx, r, masterGroup.ID, serviceDNS, ipVersion, CIDRs)
+			addRules(ctx, r, masterGroup.ID, serviceOVNDB, ipVersion, CIDRs)
+			addRules(ctx, r, masterGroup.ID, serviceKubeScheduler, ipVersion, CIDRs)
+			addRules(ctx, r, masterGroup.ID, serviceKCM, ipVersion, CIDRs)
+			addRules(ctx, r, masterGroup.ID, serviceETCD, ipVersion, CIDRs)
+			addRules(ctx, r, workerGroup.ID, serviceRouter, ipVersion, CIDRs)
+			if mastersSchedulable {
+				addRules(ctx, r, masterGroup.ID, serviceRouter, rules.EtherType4, machineV4CIDRs)
+			}
+		}
+
+		// IPv4-only rules
+		addRules(ctx, r, masterGroup.ID, serviceIKENat, rules.EtherType4, machineV4CIDRs)
+		addRules(ctx, r, workerGroup.ID, serviceIKENat, rules.EtherType4, machineV4CIDRs)
+	}()
+
+	for ruleCreateOpts := range r {
+		if _, err := rules.Create(networkClient, ruleCreateOpts).Extract(); err != nil {
+			return fmt.Errorf("failed to create the security group rule on group %q for %s %s on ports %d-%d: %w", ruleCreateOpts.SecGroupID, ruleCreateOpts.EtherType, ruleCreateOpts.Protocol, ruleCreateOpts.PortRangeMin, ruleCreateOpts.PortRangeMax, err)
+		}
+	}
```

Comment:
There's a bulk create for SG rules BTW: https://opendev.org/openstack/openstacksdk/commit/9e58f9d77df802c3d983ff8ea35a38c74eb6b2d9. No big deal we're not using it, just saying.

Reply:
Uh, I am not sure I understand. In that link I think I see SDK (ie client) code. Is it doing something different than calling the API multiple times?

I don't see any bulk creation mechanism described in [the API docs](https://docs.openstack.org/api-ref/network/v2/index.html#security-group-rules-security-group-rules). Is it an undocumented feature?

Reply:
Oh, it apparently is undocumented, but works:

```
mdulko:~/ $ openstack security group rule list test

mdulko:~/ $ curl -g -i <snip!> -d '{"security_group_rules": [{"security_group_id": "f2754a97-abd0-4538-ae66-04345dbd267a", "protocol": "tcp", "remote_ip_prefix": "192.168.0.1", "direction": "ingress", "ethertype": "IPv4"}, {"security_group_id": "f2754a97-abd0-4538-ae66-04345dbd267a", "protocol": "tcp", "remote_ip_prefix": "192.168.0.2", "direction": "ingress", "ethertype": "IPv4"}]}'
HTTP/1.1 201 Created
content-type: application/json
content-length: 1166
x-openstack-request-id: req-6ade61d0-31dc-4352-938d-ddfc04a5d3e4
date: Thu, 11 Jan 2024 11:18:15 GMT

{"security_group_rules": [{"id": "626de3f5-8048-4923-b354-60f5ca11cd9e", "tenant_id": "9e1f1eda803b4dd18ec77b3ed4d572a0", "security_group_id": "f2754a97-abd0-4538-ae66-04345dbd267a", "ethertype": "IPv4", "direction": "ingress", "protocol": "tcp", "port_range_min": null, "port_range_max": null, "remote_ip_prefix": "192.168.0.1/32", "remote_address_group_id": null, "normalized_cidr": "192.168.0.1/32", "remote_group_id": null, "description": "", "created_at": "2024-01-11T11:18:15Z", "updated_at": "2024-01-11T11:18:15Z", "revision_number": 0, "project_id": "9e1f1eda803b4dd18ec77b3ed4d572a0"}, {"id": "0da28cc5-f105-46e1-b5b3-28511aa4383d", "tenant_id": "9e1f1eda803b4dd18ec77b3ed4d572a0", "security_group_id": "f2754a97-abd0-4538-ae66-04345dbd267a", "ethertype": "IPv4", "direction": "ingress", "protocol": "tcp", "port_range_min": null, "port_range_max": null, "remote_ip_prefix": "192.168.0.2/32", "remote_address_group_id": null, "normalized_cidr": "192.168.0.2/32", "remote_group_id": null, "description": "", "created_at": "2024-01-11T11:18:15Z", "updated_at": "2024-01-11T11:18:15Z", "revision_number": 0, "project_id": "9e1f1eda803b4dd18ec77b3ed4d572a0"}]}%                                                                                                               
mdulko:~/ $ openstack security group rule list test
+--------------------------------------+-------------+-----------+----------------+------------+-----------+-----------------------+----------------------+
| ID                                   | IP Protocol | Ethertype | IP Range       | Port Range | Direction | Remote Security Group | Remote Address Group |
+--------------------------------------+-------------+-----------+----------------+------------+-----------+-----------------------+----------------------+
| 0da28cc5-f105-46e1-b5b3-28511aa4383d | tcp         | IPv4      | 192.168.0.2/32 |            | ingress   | None                  | None                 |
| 626de3f5-8048-4923-b354-60f5ca11cd9e | tcp         | IPv4      | 192.168.0.1/32 |            | ingress   | None                  | None                 |
+--------------------------------------+-------------+-----------+----------------+------------+-----------+-----------------------+----------------------+
```

Reply:
https://review.opendev.org/c/openstack/neutron-lib/+/905299

Reply:
I'd rather not instrument an undocumented call into Gophercloud. Thanks for the heads-up though, as I had no idea such call existed

Reply:
(would be cool to add a remote-prefix field in the example while you're at it, as they're completely missing in he existing examples)

Reply:
Rodolfo says it's been there for a while. ;)

> It has been there always (11 years): https://review.opendev.org/c/openstack/neutron/+/14262

Reply:
https://docs.openstack.org/api-ref/network/v2/index.html#bulk-create-security-group-rule - and it's documented now. ;)